src/gg20/crypto_tools/paillier/zk/mod.rs

```diff
@@ -51,17 +55,50 @@ impl ZkSetup {
         Ok(Self::from_keypair(rng, keypair, domain))
     }
 
+    /// Add a layer of domain separation on the two composite dlog proofs
+    fn compute_domain(domain: &[u8]) -> (Vec<u8>, Vec<u8>) {
+        let mut domain1: Vec<u8> = domain.into();
+        let mut domain2: Vec<u8> = domain.into();
+
+        domain1.push(constants::COMPOSITE_DLOG_PROOF1);
+        domain2.push(constants::COMPOSITE_DLOG_PROOF2);
+
+        (domain1, domain2)
+    }
+
     fn from_keypair(
         rng: &mut (impl CryptoRng + RngCore),
         (ek_tilde, dk_tilde): (EncryptionKey, DecryptionKey),
         domain: &[u8],
     ) -> (ZkSetup, ZkSetupProof) {
-        let (dlog_stmt, witness) =
-            CompositeDLogStmt::setup(rng, ek_tilde.0.n(), dk_tilde.0.p(), dk_tilde.0.q());
-
-        let dlog_proof = dlog_stmt.prove(&witness, domain);
-
-        (Self { dlog_stmt }, dlog_proof)
+        let (dlog_stmt1, witness1, witness2) = CompositeDLogStmt::setup(
+            rng,
+            ek_tilde.0.n(),
+            dk_tilde.0.p(),
+            dk_tilde.0.q(),
+            dk_tilde.0.totient(),
+        );
+
+        let dlog_stmt2 = CompositeDLogStmt {
+            n: dlog_stmt1.n.clone(),
+            g: dlog_stmt1.v.clone(),
+            v: dlog_stmt1.g.clone(),
+        };
```

Comment:
This is a usability gotcha.  In order to use `witness2` you need to know to compose your own `stmt2` from the `stmt1` you have by swapping `g`, `v`.  At the very least we need to document this.  But perhaps a better solution is to have `CompositeDLogStmt::setup` just return `(stmt1, witness1, stmt2, witness2)`.

This return tuple is getting large though.  Perhaps it's time to bundle these statements/witnesses into a larger struct?

Reply:
Yeah, I'll create a method in `CompositeDLogStmt` instead, since we need to do this in `verify` too, where `::setup` isn't being called.
Alternatively, we could modify the zk proof to prove two things internally, so the proof itself will have `y1` and `y2` elements (this would also save some time by not rechecking everything again). The downside is that the zk proof is not as standalone and doesn't simply mimic the paper.

Reply:
I like the new `get_inverse_statement` method.